certbot-apache/certbot_apache/interfaces.py

```diff
@@ -0,0 +1,281 @@
+    by marking modified ParserNodes as dirty to enable partial write-to-disk for
+    different files in the configuration structure.
+
+    While for the most parts the usage and the child types are obvious, "include"-
+    and similar directives are an exception to this rule. This is because of the
+    nature of include directives - which unroll the contents of another file or
+    configuration block to their place. While we could unroll the included nodes
+    to the parent tree, it remains important to keep the context of include nodes
+    separate in order to write back the original configuration as it was.
+    """
+
+    @property
+    @abc.abstractmethod
+    def ancestor(self):
+        """
+        This property contains a reference to ancestor node, or None if the node
+        is the root node of the configuration tree.
+        """
+
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def arguments(self):
+        """
+        This property contains a list of arguments of this ParserNode object.
+        """
+
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def dirty(self):
+        """
+        This property contains a boolean value of the information if this node has
+        been modified since last save (or after the initial parse).
+        """
+
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def save(self, msg):
+        """
+        Save traverses the children, and attempts to write the AST to disk for
+        all the objects that are marked dirty. The actual operation of course
+        depends on the underlying implementation. save() shouldn't be called
+        from the Configurator outside of it's designated save() method in order
+        to ensure that the Reverter checkpoints are created properly.
+
+        Note: this approach of keeping internal structure of the configuration
+        within the ParserNode tree does not represent the file inclusion structure
+        of actual configuration files that reside in the filesystem. To handle
+        file writes properly, the file specific temporary trees should be extracted
+        from the full ParserNode tree where necessary when writing to disk.
+
+        """
+
+    @abc.abstractmethod
+    def set_arguments(self, arguments):
+        """
+        Sets argument list of the ParserNode object and marks the node dirty.
+        """
+
+
+@six.add_metaclass(abc.ABCMeta)
+class CommentNode(ParserNode):
+    """
+    CommentNode class is used for representation of comments within the parsed
+    configuration structure. Because of the nature of comments, it is not able
+    to have child nodes and hence it is always treated as a leaf node.
+
+    CommentNode stores its contents in class variable 'arguments' and does not
+    have a specific name.
+    """
+
+    @property
+    @abc.abstractmethod
+    def comment(self):
```

Comment:
Is this different from `arguments?

Reply:
It's basically the same, but returning a string instead of list. I find this more natural for comments. If you disagree, we can definitely just handle the contents as arguments here too.

Reply:
What do you think? If we want to use `arguments` property instead, especially with the current plan of writing a whitespace-hiding list extension, I feel that `comment` would suit it better. If we decide to do this, I'll move the `arguments` property to `DirectiveNode` and `BlockNode` interfaces instead of the `ParserNode`.

Reply:
The plan in your most recent comment SGTM!

My only question becomes how do we update the comment?

Proposal: Allow the user to change the `comment` attribute directly. This should be pretty easy to implement by giving the comment a setter property (which we maybe can/should define in the ABC here for clarity?) so we can set the dirty bit when the comment is modified.